File: PyTorch/arch/splitNetworks.py
"""
 Splitting trained single-encoder-multi-decoder into three networks
 Copyright (c) 2019 Samsung Electronics Co., Ltd. All Rights Reserved
 If you use this code, please cite the following paper:
 Mahmoud Afifi and Michael S Brown. Deep White-Balance Editing. In CVPR, 2020.
"""
__author__ = "Mahmoud Afifi"
__credits__ = ["Mahmoud Afifi"]

import copy

from arch import deep_wb_single_task


def splitNetworks(net):
    # create instances from deepWBnet (the single task model)
    net_awb = deep_wb_single_task.deepWBnet()
    net_t = deep_wb_single_task.deepWBnet()
    net_s = deep_wb_single_task.deepWBnet()
    # copy AWB weights
    net_awb.encoder_inc = copy.deepcopy(net.encoder_inc)
    net_awb.encoder_down1 = copy.deepcopy(net.encoder_down1)
    net_awb.encoder_down2 = copy.deepcopy(net.encoder_down2)
    net_awb.encoder_down3 = copy.deepcopy(net.encoder_down3)
    net_awb.encoder_bridge_down = copy.deepcopy(net.encoder_bridge_down)
    net_awb.decoder_bridge_up = copy.deepcopy(net.awb_decoder_bridge_up)
    net_awb.decoder_up1 = copy.deepcopy(net.awb_decoder_up1)
    net_awb.decoder_up2 = copy.deepcopy(net.awb_decoder_up2)
    net_awb.decoder_up3 = copy.deepcopy(net.awb_decoder_up3)
    net_awb.decoder_out = copy.deepcopy(net.awb_decoder_out)
    # copy Tungsten WB weights
    net_t.encoder_inc = copy.deepcopy(net.encoder_inc)
    net_t.encoder_down1 = copy.deepcopy(net.encoder_down1)
    net_t.encoder_down2 = copy.deepcopy(net.encoder_down2)
    net_t.encoder_down3 = copy.deepcopy(net.encoder_down3)
    net_t.encoder_bridge_down = copy.deepcopy(net.encoder_bridge_down)
    net_t.decoder_bridge_up = copy.deepcopy(net.tungsten_decoder_bridge_up)
    net_t.decoder_up1 = copy.deepcopy(net.tungsten_decoder_up1)
    net_t.decoder_up2 = copy.deepcopy(net.tungsten_decoder_up2)
    net_t.decoder_up3 = copy.deepcopy(net.tungsten_decoder_up3)
    net_t.decoder_out = copy.deepcopy(net.tungsten_decoder_out)
    # copy Shade WB weights
    net_s.encoder_inc = copy.deepcopy(net.encoder_inc)
    net_s.encoder_down1 = copy.deepcopy(net.encoder_down1)
    net_s.encoder_down2 = copy.deepcopy(net.encoder_down2)
    net_s.encoder_down3 = copy.deepcopy(net.encoder_down3)
    net_s.encoder_bridge_down = copy.deepcopy(net.encoder_bridge_down)
    net_s.decoder_bridge_up = copy.deepcopy(net.shade_decoder_bridge_up)
    net_s.decoder_up1 = copy.deepcopy(net.shade_decoder_up1)
    net_s.decoder_up2 = copy.deepcopy(net.shade_decoder_up2)
    net_s.decoder_up3 = copy.deepcopy(net.shade_decoder_up3)
    net_s.decoder_out = copy.deepcopy(net.shade_decoder_out)

    return net_awb, net_t, net_s

File: Matlab/arch/splitNetworks.m
%% Splitting trained single-encoder-multi-decoder into three networks 
% Author: Mahmoud Afifi
% Copyright (c) 2019 Samsung Electronics Co., Ltd. All Rights Reserved
% Please cite our paper:
% Mahmoud Afifi and Michael S Brown. Deep White-Balance Editing. In CVPR, 2020.
%%

function [AWB_net,T_net,S_net]  = splitNetworks(net, depth)

net = layerGraph(net);

AWB_net = net;
T_net = net;
S_net = net;

clear net;

prefixs= {'T-','S-'};
for p = 1 : length(prefixs)
    AWB_net = removeLayers(AWB_net,{AWB_net.Layers(...
        contains({AWB_net.Layers(:).Name},prefixs(p))).Name});
end

cutindx = find(strcmp({AWB_net.Layers(:).Name},...
    sprintf('Encoder-Stage-%d-MaxPool',depth)));

AWBDec_layer_names = {AWB_net.Layers(cutindx+2:end-1).Name};

T_net = removeLayers(T_net,AWBDec_layer_names);
S_net = removeLayers(S_net,AWBDec_layer_names);

for p = 1 : length(prefixs)
    modelName = prefixs{p};
    modelName = [modelName(1) '_net'];
    curr_prefixs = setdiff(prefixs,prefixs{p});
    for cp = 1 : length(curr_prefixs)
        
        eval(sprintf('%s=removeLayers(%s,{%s.Layers(contains({%s.Layers(:).Name},curr_prefixs{%d})).Name});',...
            modelName,modelName,modelName,modelName,cp));
        
    end
end
T_net = removeLayers(T_net,'FinalLayer');
S_net = removeLayers(S_net,'FinalLayer');
AWB_net = removeLayers( AWB_net,'catLayer');
AWB_net = removeLayers( AWB_net,'outLayer');
AWB_net = removeLayers( AWB_net,'FinalLayer');


T_net = dlnetwork(T_net);
S_net = dlnetwork(S_net);
AWB_net = dlnetwork(AWB_net);